src/PPrintEngine.ml

```diff
@@ -58,39 +58,90 @@ class type output = object
 
 end
 
+(* ------------------------------------------------------------------------- *)
+
+(* Printing blank space (indentation characters). *)
+
+let blank_length =
+  80
+
+let blank_buffer =
+  String.make blank_length ' '
+
+let rec blanks output_fn n =
+  if n <= 0 then
+    ()
+  else if n <= blank_length then
+    output_fn blank_buffer 0 n
+  else begin
+    output_fn blank_buffer 0 blank_length;
+    blanks output_fn (n - blank_length)
+  end
+
+(* ------------------------------------------------------------------------- *)
+
+class with_blank_buffering char substring = object(self)
+  val mutable blank_buffered = None 
+
+  method private flush_blank =
+    match blank_buffered with
+    | None -> ()
+    | Some n -> 
+      blanks substring n;
+      blank_buffered <- None
+
+  method char c : unit =
+    begin match c with
+    | '\n' -> blank_buffered <- Some 0
+    | _ -> self#flush_blank
+    end;
+    char c
+
+  method substring s pos len =
+    if s == blank_buffer then (
```

Comment:
The physical comparison here seems a bit fragile and obscure. It would be good to document what is going on here, or to find a less obscure way of doing this.